### PR TITLE
chore: Port `ctl` code to Typescript

### PR DESCRIPTION
**Pull request by sharat87**
This PR converts the Javscript code of the `ctl` module, into Typescript, and in the process already fixing two small bugs that went undetected because... of lack of good type checking.

The linting exceptions are still there and will be removed in the next PR. Didn't want to change anymore than necessary or Git will detect these changes as "new files" instead of as "renames".

## Automation

/test sanity

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11949059369>
> Commit: e156dacbc5cb513030052535c3d1f25ce1c7f222
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11949059369&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Sanity`
> Spec:
> <hr>Thu, 21 Nov 2024 08:45:48 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Release Notes

- **New Features**
  - Introduced TypeScript support by updating entry points and adding type definitions for `nodemailer` and `readline-sync`.
  - Enhanced logging and error handling functionalities across various modules.

- **Improvements**
  - Transitioned from CommonJS to ES module syntax for better compatibility and maintainability.
  - Improved clarity and structure in command handling and test files.

- **Bug Fixes**
  - Corrected regex patterns and variable declarations in tests to enhance reliability.

- **Chores**
  - Updated dependencies and refined module exports for better organization.

### CODE REVIEW

File: app/client/packages/rts/src/ctl/backup.test.ts [hunk 1]
```diff
@@ -1,15 +1,20 @@
...
-const os = require("os");
-const fsPromises = require("fs/promises");
-const utils = require("./utils");
-const readlineSync = require("readline-sync");
+jest.mock("./utils", () => ({
+  ...jest.requireActual("./utils"),
+  execCommand: jest.fn().mockImplementation(async (a) => a.join(" ")),
+}));
+
+import * as backup from "./backup";
```
Comment (trishaanand) [inline app/client/packages/rts/src/ctl/backup.test.ts:6]: Do we do * imports on the client side code base? Should we change this to particular func exports instead?

File: app/client/packages/rts/src/ctl/mailer.ts [hunk 2]
```diff
@@ -77,7 +77,7 @@ async function sendBackupErrorToAdmins(err, backupTimestamp) {
           user: mailUser,
           pass: mailPass,
         },
-      });
+      } as any);
```
Comment (trishaanand) [inline app/client/packages/rts/src/ctl/mailer.ts:80]: Doesn't any surpass the type checks?

**Review by trishaanand — APPROVED** [review]
Minor comments added. LGTM otherwise.

File: app/client/packages/rts/src/ctl/mailer.ts (hunk 2 above)
Comment (sharat87) [reply app/client/packages/rts/src/ctl/mailer.ts:80]: Yes, they do. That's the intention here. There's one other place I had to do this.

Why? One, I don't want to change the files _too_ much, or Git won't treat it as a rename, which I really _really_ want. So I want to fix this in another PR.

Two, the type definitions for this library don't think `host` should be provided. Which is very weird, but the code clearly has been working. I have to check if the types are incorrect, or is our usage has always been incorrect.

**Comment by sharat87** [comment]
Deferring CodeRabbit suggestions to the next PR so I don't change the files _too_ much. I want Git to treat the changes as file _renames_ instead of deletions+additions.

**Pull request merged**